crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,753 @@
+                                 OSSL_CRMF_SINGLEPUBINFO *spi)
+{
+    if (pi == NULL || spi == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_PKIPUBLICATIONINFO_PUSH0_SINGLEPUBINFO,
+                CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (pi->pubinfos == NULL)
+        if ((pi->pubinfos = sk_OSSL_CRMF_SINGLEPUBINFO_new_null()) == NULL)
+            goto oom;
+
+    if (!(sk_OSSL_CRMF_SINGLEPUBINFO_push(pi->pubinfos, spi)))
+        goto oom;
+    return 1;
+
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PKIPUBLICATIONINFO_PUSH0_SINGLEPUBINFO,
+            ERR_R_MALLOC_FAILURE);
+    return 0;
+}
+
+int OSSL_CRMF_MSG_set_PKIPublicationInfo_action(
+                                 OSSL_CRMF_PKIPUBLICATIONINFO *pi, int action)
+{
+    if (pi == NULL
+        || action < OSSL_CRMF_PUB_ACTION_DONTPUBLISH
+        || action > OSSL_CRMF_PUB_ACTION_PLEASEPUBLISH) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_PKIPUBLICATIONINFO_ACTION,
+                ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+
+    return ASN1_INTEGER_set(pi->action, action);
+}
+
+ /* id-regCtrl-pkiPublicationInfo Control (section 6.3) */
+IMPLEMENT_CRMF_CTRL_FUNC(pkiPublicationInfo, OSSL_CRMF_PKIPUBLICATIONINFO,
+                         regCtrl)
+
+ /* id-regCtrl-oldCertID Control (section 6.5) from the given */
+IMPLEMENT_CRMF_CTRL_FUNC(oldCertID, OSSL_CRMF_CERTID, regCtrl)
+
+OSSL_CRMF_CERTID *OSSL_CRMF_CERTID_gen(const X509_NAME *issuer,
+                                       const ASN1_INTEGER *serial)
+{
+    OSSL_CRMF_CERTID *cid = NULL;
+
+    if (issuer == NULL || serial == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_CERTID_GEN, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+
+    if ((cid = OSSL_CRMF_CERTID_new()) == NULL)
+        goto oom;
+
+    if (!X509_NAME_set(&cid->issuer->d.directoryName, issuer))
+        goto oom;
+    cid->issuer->type = GEN_DIRNAME;
+
+    ASN1_INTEGER_free(cid->serialNumber);
+    if ((cid->serialNumber = ASN1_INTEGER_dup(serial)) == NULL)
+        goto oom;
+
+    return cid;
+
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_CERTID_GEN, ERR_R_MALLOC_FAILURE);
+    OSSL_CRMF_CERTID_free(cid);
+    return NULL;
+}
+
+ /* id-regCtrl-protocolEncrKey Control (section 6.6) */
+ /*
+  * For some reason X509_PUBKEY_dup() is not implemented in OpenSSL X509
+  * TODO: check whether that should go elsewhere
+  */
+static IMPLEMENT_ASN1_DUP_FUNCTION(X509_PUBKEY)
+IMPLEMENT_CRMF_CTRL_FUNC(protocolEncrKey, X509_PUBKEY, regCtrl)
+
+/*
+ * Pushes the attribute given in regInfo in to the CertReqMsg->regInfo stack.
+ * (section 7)
+ * returns 1 on success, 0 on error
+ */
+static int OSSL_CRMF_MSG_push0_regInfo(OSSL_CRMF_MSG *crm,
+                                       OSSL_CRMF_ATTRIBUTETYPEANDVALUE *ri)
+{
+    int new = 0;
+
+    if (crm == NULL || ri == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_REGINFO, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (crm->regInfo == NULL) {
+        if ((crm->regInfo =
```

Comment:
we have other places where we do it this way:
```C
OSSL_CRM_MSG *info = NULL;

if (crm->regInfo == NULL)
    crm->regInfo = info = xxx_new();

if (crm->regInfo == NULL)
    goto oom;

...
oom:
   if (info != NULL)
       crm->regInfo = NULL;
   xxx_free(info);
```

Reply:
Nice pattern; changed like this.